### ISSUE TITLE
More use of FastUtil

### ISSUE THREAD

**Issue by andreasdc** [ProjectKig/KigPaper#9]
What do you think about changing server's collections? Like player list or other collections, to the ones that have better performance? That would need heavy testing and benchmarks of course.
I saw Paper uses fastutil in their version of PlayerChunkMap, I don't remember exact name of the new class, but I think it's PlayerChunk, it really looks like it.
There is fastutil for example in:
PacketPlayOutEntityDestroy
PacketPlayOutGameStateChange
PacketPlayOutMultiBlockChange
PlayerMap
NBTTagLongArray
Chunk
NBTTagList

I think fastutil in PlayerMap (PlayerChunkMap and dirtyBlocks on 1.8) will be a huge improvement. What do you think?

**Comment by archer-321** [comment]
As you say, this would need some benchmarking, but it might decrease the ms/tick quite considerably.
I most likely won't have time to test it in the next couple of days, but it's worth a try for sure.

**Comment by andreasdc** [comment]
> As you say, this would need some benchmarking, but it might decrease the ms/tick quite considerably. I most likely won't have time to test it in the next couple of days, but it's worth a try for sure.

Yes but Paper already use fastutil in for example classes that I mentioned, so I think we could trust them in that.

**Comment by andreasdc** [comment]
Bump?

**Comment by archer-321** [comment]
Currently, we don't have any problems with collections in KigPaper. The server's collections aren't a bottleneck in 1.8.8, and as far as we can see, the effort of benchmarking different implementations, replacing the existing collections, and maintaining the new collections outweighs the potential performance increase.

If you're willing to test different collection implementations, we'll gladly review a pull request, but currently, we're not going to investigate different collection implementations any further.

**Comment by andreasdc** [comment]
> Currently, we don't have any problems with collections in KigPaper. The server's collections aren't a bottleneck in 1.8.8, and as far as we can see, the effort of benchmarking different implementations, replacing the existing collections, and maintaining the new collections outweighs the potential performance increase.
> 
> If you're willing to test different collection implementations, we'll gladly review a pull request, but currently, we're not going to investigate different collection implementations any further.

I think the best will be in PlayerChunkMap, they have fastutil ShortSets on 1.17, every block change etc. has to go through it.
PlayerChunkMap 1.8:
https://pastebin.com/raw/s3W0Amk6
PlayerMap 1.17:
https://pastebin.com/raw/DmFDyLz2

Additionally I can say that changes in PlayerChunkMap cost more than actually changing blocks.
